pkg/resource/claim_binding_reconciler.go

```diff
@@ -394,6 +382,14 @@ func (r *ClaimReconciler) Reconcile(req reconcile.Request) (reconcile.Result, er
 			return reconcile.Result{RequeueAfter: aShortWait}, errors.Wrap(r.client.Status().Update(ctx, claim), errUpdateClaimStatus)
 		}
 
+		if err := r.claim.AddFinalizer(ctx, claim); err != nil {
```

Comment:
This would add finalizer to claim only if the managed resource is not created yet. I know that we have scenarios where the managed resource exists and our newly created claim binds to it, so, in those cases we'd see an immediate deletion of the claim. What do you think about calling `r.claim.AddFinalizer ` right before `if !meta.WasCreated(managed) && claim.GetClassReference() != nil` just like managed reconciler?

Reply:
Now that you mention it it's likely that this is why statically provisioned managed resources are never properly unbound per https://github.com/crossplaneio/crossplane-runtime/issues/25. I agree that this isn't the right place to add the finalizer, but I actually think we should set it right before binding to our managed resource.

The managed resource reconciler uses the finalizer to delete the external resource (if necessary) whereas the claim reconciler uses the finalizer to _unbind_ the managed resource. If we never bound to the resource there's nothing to do. Keep in mind we currently rely on garbage collection to ensure dynamically provisioned managed resources are deleted when the claim is deleted.

Reply:
> the claim reconciler uses the finalizer to unbind the managed resource

> we currently rely on garbage collection to ensure dynamically provisioned managed resources are deleted when the claim is deleted.

Well, aren't these anti-pattern anyway? Bind/Unbind mechanism should only be meaningful if the reclaim policy between claim and managed resource is set to `Retain`. If it's `Delete`, we should just delete it explicitly. But there is no claim<>managed reclaim policy and all this seems to be related to #21 , so, it should be OK for this PR not to touch that.

IMHO, sounds to me it's time we finally fix #21 and #27 . I think I can work on a one-pager that covers both. Not much to talk about #21 as it will just work like PVC/PV model but I think it'd be good to cover both in the same document to explain the mechanisms better. Maybe after completing Azure redis & db v1beta1. @prasek @negz what do you guys think?

Reply:
> Well, aren't these anti-pattern anyway?

It is. I agree that we will likely need to refactor how we add and remove claim finalizers once we're not relying on garbage collection to delete managed resources when the claim is deleted.

> IMHO, sounds to me it's time we finally fix #21 and #27 . I think I can work on a one-pager that covers both. Not much to talk about #21 as it will just work like PVC/PV model but I think it'd be good to cover both in the same document to explain the mechanisms better.

I can see your point, but keep in mind @hasheddan already produced a one pager for #21. I agree that design done for #27 should consider how it relates to reclaim policies but let's be careful not to re-do that work if we don't need to.

> Maybe after completing Azure redis & db v1beta1. @prasek @negz what do you guys think?

My suspicion is that #21 will mostly involve changes to the resource claim reconciler, while #27 could impact how we write `resource.ExternalClient` implementations. On one hand I can see that these are both potentially "it's never the right time" kind of issues, but I feel like they're lower priority than getting all of our managed resource controllers ported to the managed resource reconciler (not necessarily v1beta1).